CAMEL-11261 Revise Camel context destruction in Spring (Boot) applications

zregvart opened the pull request:
Submitted for review, thanks 👍 

See the discussion on the [user forum](http://camel.465427.n5.nabble.com/OnApplicationEvent-not-called-with-CamelAutoConfiguration-tt5798802.html).

This makes `CamelSpringBootApplicationController` lifecycle bean with
top priority during `ApplicationContext` close.

nicolaferraro commented:
I did not try your changes, sorry! I just did some debugging on the current shutdown process, but I think the application controller is not responsible for shutting down the Camel context (in spring-boot), but just to keep it alive ... Even if it actually closes the Camel context, the "destroy()" method of the Camel Context is *also* called by Spring automagically.

My concern is that if you don't put `camel.springboot.main-run-controller=true` (e.g. when you use `spring-boot-starter-web` spring creates other non-daemon threads), the application controller is not run at all, even if it's created. So your changes should have no effect in that case.
There's *also* a shutdown hook in MainSupport (always active, I think we should not use it in spring-boot) that closes the application controller outside of the spring lifecycle...

So, it seems that the camel context gets closed three times...

We should make sure that the Camel context is closed before the managed connection factory in any case... What about using the same approach with the CamelContext object?

zregvart commented:
@nicolaferraro yeah, that was what I thought but [CamelAutoConfiguration.java#L203-L211](https://github.com/apache/camel/blob/master/components/camel-spring-boot/src/main/java/org/apache/camel/spring/boot/CamelAutoConfiguration.java#L203-L211) creates `CamelSpringBootApplicationController` as a `@Bean` regardless, and I thought that this would be a good place to put Lifecycle. But I do agree that we should have a holistic approach that works in both Spring Boot and in Spring framework. So perhaps implementing this on `SpringCamelContext`?

zregvart commented:
@nicolaferraro could you take another look, couple of things to note:
- I've removed `DisposableBean` implement on `SpringCamelContext` the destruction should now be handled by lifecycle
- in `CamelAutoConfiguration` I've explicitly set `destroyMethod` to `""` so that it is not inferred as Spring magically recognizes `shutdown` (see [Bean.java#L239-L250](https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/context/annotation/Bean.java#L239-L250))
- the use of `sneakyThrow` to throw checked exception without declaring it to maintain compatibility (see [CircularComponentCreationTest.java#L37-L38](https://github.com/apache/camel/blob/master/components/camel-spring/src/test/java/org/apache/camel/spring/CircularComponentCreationTest.java#L37-L38))

zregvart commented:
I've created JIRA issue [CAMEL-11261](https://issues.apache.org/jira/browse/CAMEL-11261)

zregvart commented:
Should `SpringCamelContext` implement the [SmartLifecycle](https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/context/SmartLifecycle.java) instead?

davsclaus commented:
Yeah maybe that SmartLifecycle allows us to set a value of the phase so CamelContext gets stopped as one of the first, then it can do its graceful shutdown while connection pools and database resources are still alive and open.

zregvart commented:
@davsclaus the PR makes `SpringCamelContext` implement `Phased` with maximum priority (https://github.com/apache/camel/pull/1685/files#diff-d7c765d52f03353dc1bb44391c81464dR306) so as it stands with this `SpringCamelContext` should be first in line to get stopped. The advantage of `SmartLifecycle` is that the `stop` can be done asynchronously, for instance `SpringCamelContext::stop` can proceed on the calling thread and the given `Runnable` can be executed on another (_or vice versa_).
Not sure if this would be a cause of pain in the future, but it could help by making the shutdown a bit faster for the user.

davsclaus commented on components/camel-spring/src/main/java/org/apache/camel/spring/SpringCamelContext.java:
```diff
@@ -301,4 +301,36 @@ public String toString() {
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <E extends Exception> void sneakyThrow(Exception e) throws E {
```
Please use the existing API we have to wrap as runtime exception

    throw ObjectHelper.wrapRuntimeCamelException(e);

davsclaus commented on components/camel-spring-boot/src/main/java/org/apache/camel/spring/boot/CamelAutoConfiguration.java:
```diff
@@ -77,7 +77,7 @@
      * Spring-aware Camel context for the application. Auto-detects and loads all routes available in the Spring context.
      */
-    @Bean
+    @Bean(destroyMethod = "")
```
We may have done this on some other places AFAIR, maybe look if that is the case and see if we did not add some code comments why we do this, as its confusing why that string is empty

nicolaferraro commented:
It makes sense to use SmartLifecycle if we have an asynchronous shutdown process, but I remember we have just a blocking one. They should be equivalent anyway (semantically).

davsclaus commented:
You need to let Camel graceful shutdown and process all its inflight messages before you can continue and let other spring managed beans shutdown. I cannot see how you can make that async and concurrent. Only after all the inflight messages has been processed you may be able to let spring continue and then let camel shutdown its remaining internal resources but they ought to be quick. eg its the inflight that may take up time.

nicolaferraro commented:
Yes, anyway shutdown with SmartLifecycle should be asynchronous but respecting the order of beans with different priority (I.e. if you don't call the callback it doesn't proceed). That's why I said it's semantically equivalent, even if it doesn't provide any real advantage in our case.. it's just a matter of having a thread in blocked state, vs not having it.. which is not so different since we are shutting down the app..

zregvart commented on components/camel-spring-boot/src/main/java/org/apache/camel/spring/boot/TypeConversionConfiguration.java:
```diff
@@ -33,14 +33,18 @@
 
-    @Bean(initMethod = "", destroyMethod = "")
-    // Camel handles the lifecycle of this bean
+    @Bean
```
I've removed `initMethod` and `destroyMethod` as `SpringTypeConverter` is a concrete class and does not provide `stop` or `shutdown` methods that might be auto-configured for bean destruction.

zregvart commented:
@nicolaferraro and @davsclaus thanks for the feedback 👍 I think this would be ready now to merge upstream, but I'll leave it if you want to take another look.

davsclaus commented:
Have you run camel-spring and camel-spring-boot unit tests to see if any odd test may fail? Also a good idea is to run the tests in tests/camel-itest which has a bunch of different tests, and some are spring related.

Also if you have the time running the tests in the example may sometimes catch mistakes.

zregvart commented:
Yeah, I've run the tests in `camel-spring` and `camel-spring-boot`, but I'll run a `mvn clean install` from root just to make sure.

davsclaus commented:
Thanks, you can leave it overnight. I had mine run for 6h about the error handler thingy.

We have in the past got bitten when changed some lifecycle stuff in either spring or blueprint, which would cause side-effects / errors in surprising places or that some obscure tests in itest such as eg those cxf blueprint tests etc. Or in one of the examples where for example an AMQ broker is embedded etc.

But yeah looks good from my pov the PR if it makes shutdown better for spring-boot. And you can tell how old Camel is as it was before that SmartLifecycle ;)

zregvart commented:
@davsclaus you might need a new computer 😃 on mine it took ⏩ 03:49

    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD SUCCESS
    [INFO] ------------------------------------------------------------------------
    [INFO] Total time: 03:49 h

The joke aside, I'll run the examples and if everything is OK, merge this to master

nicolaferraro commented:
@zregvart it looks very good! I hope that the tests are passing.

nicolaferraro commented:
Indeed they are :)

zregvart commented:
There is a slight quirk with this approach, the `SpringCamelContext` created by `CamelContextFactoryBean` will not be destroyed (stopped) using the lifecycle workflow, but using the event workflow, meaning when `ApplicationListener` receives `ContextClosedEvent`. This is because the `ContextClosedEvent` is published before lifecycle (see [AbstractApplicationContext.java#L993](https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/context/support/AbstractApplicationContext.java#L992)) and the `DefaultLifecycleProcessor` checks if the `FactoryBean` implements the `Lifecycle`, not the bean it produced. The `CamelContext` will be closed but not in the same workflow as in Spring Boot.

So, if we want go on the `Lifecycle` route the `CamelContextFactoryBean` needs to implement `Lifecycle` also. I can update this to include that and to remove handling of `ContextClosedEvent` and `ContextStoppedEvent` from `SpringCamelContext::onApplicationEvent`.

Or we can backtrack and handle `ContextClosedEvent` in `CamelContext` created by Spring Boot autoconfiguration. This perhaps might be a better option, because there could be other Spring beans listening for `ContextClosedEvent` and reacting to it before the lifecycle workflow. To jump ahead of other Spring beans in the event shutdown workflow `SpringCamelContext` would need to implement `ApplicationListener` (it does not now, even as it has `onApplicationEvent(ApplicationEvent)` method) and `Ordered` (with top priority), and we probably need to take another look at how the responsibilities are divided between `CamelContextFactoryBean`/Spring Boot auto configuration and `SpringCamelContext`.

What do you think?

nicolaferraro commented:
I think we should try to use the `Lifecycle` API to implement initialization/disposal, even if other (hacky?) approaches may lead to better results. I have also seen problems in the thread linked to this PR: https://issues.apache.org/jira/browse/CAMEL-2607. Don't understand exactly what were happening at the end, but the current implementation seems a workaround around that issue.

Using `Lifecycle` also on the factory may be better. @davsclaus should have a better historical perspective on this.

p.s. you're right to be cautious for this issue, the risk of breaking something is really high :+1:

kudos to Zoran!

davsclaus commented:
btw can we run the spring-boot itests before merging this PR so we know its state beforehand? The CI server had a failure in the servlet test but it may have been due other things, but would be good just to know the before vs after as its spring-boot after all.

zregvart commented:
We have [CamelSpringBootShutdownTest](https://github.com/apache/camel/blob/master/components/camel-spring-boot/src/test/java/org/apache/camel/spring/boot/CamelSpringBootShutdownTest.java) but I think it touches upon just one aspect of this. So I'll create startup/shutdown integration tests for `camel-spring-boot` and `camel-spring`, in the later one when using `CamelContextFactoryBean` and one when using `SpringCamelContext`.

nicolaferraro commented:
@davsclaus which failure are you referring to? I see every time different tests failing here: https://builds.apache.org/view/A-D/view/Camel/job/camel.trunk.itest.springboot/ (but not servlet).

davsclaus commented:
@nicolaferraro I think the CI server only keep 3 jobs back or thereabouts so the job is likely gone now.

zregvart commented:
@nicolaferraro, @davsclaus I got this to pass tests in `camel-spring` and `camel-spring-boot` I'm running tests on the whole project just to make sure. I had to change some of the tests as now at different points different exceptions are thrown than the previously expected ones. There is a new test in `camel-spring-boot` that tests the startup/shutdown order, I think I made the right hypothesis but could you have a look at that ([StartupShutdownOrderTest.java](https://github.com/zregvart/camel/blob/dd57302f4849a295f0ca375dc30f67c108f3047e/components/camel-spring-boot/src/test/java/org/apache/camel/spring/boot/StartupShutdownOrderTest.java)) as well. I would like to also test with one of the examples and try to reproduce the issue with current _master_ and with this branch to see if it works in real world, not just in integration tests :)

Thank you for taking a look 🥇

zregvart commented:
There are failures in `camel-test-spring`, I'm looking into those

zregvart commented:
I've split this into three different commits to help with the review:
 1. [Hypothesis](https://github.com/apache/camel/pull/1685/commits/456b3fdede27c5bba02616596d52fd9b45ae6761)
 2. [Code changes](https://github.com/apache/camel/pull/1685/commits/6d508d7f071b10ff6da5a3398cc1e44d9f9d9e46)
 3. [Unit test fixes/cleanup](https://github.com/apache/camel/pull/1685/commits/933fb8c7eb90e85067488bfb83d43372854cf1c0)

In short, I've done away with `SpringCamelContext.setNoStart(true)` way of initialization and used `SmartLifecycle`, explicitly declared `@Bean(destroyMethod = "")` as Spring would pick up `stop` and `shutdown` methods auto-magically. Revised the tests where this would cause an issue, and changed the tests to assert the newly thrown exception hierarchy.

@nicolaferraro, @davsclaus could you take another look, I think this is a big change and I don't want to miss anything. Thanks 🥇

davsclaus commented on components/camel-spring/src/main/java/org/apache/camel/spring/CamelContextFactoryBean.java:
```diff
@@ -341,43 +338,49 @@ protected void initPropertyPlaceholder() throws Exception {
 
+    @Override
+    public void stop(Runnable callback) {
+        stop();
```
Should the callback always be called, eg if stop throw an exception should we put it in a finally block?

zregvart replied:
Yeah it will end with a [warning in the log file](https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/context/support/DefaultLifecycleProcessor.java#L259), and I guess that the [default logic in Spring](https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/context/support/DefaultLifecycleProcessor.java#L374-L378) depends on not calling the runnable in case of exceptions. The [javadoc](http://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/SmartLifecycle.html#stop-java.lang.Runnable-) unfortunately does not mention an error scenario.

davsclaus commented on components/camel-spring/src/main/java/org/apache/camel/spring/SpringCamelContext.java:
```diff
@@ -266,39 +233,58 @@ protected ModelJAXBContextFactory createModelJAXBContextFactory() {
+    }
+
+    @Override
+    public int getPhase() {
```
This one has MAX value and the other had MIN value

zregvart replied:
I think I need to do more tests, I did this in a TDD style and arrived here, but I most likely have missed some scenarios. Having those different is a smell...

davsclaus commented:
Looks really good, the only comment is the MAX vs MIN which likely could use some code comments why one is this and the other that. And that callback thing I also commented one.

zregvart commented:
So, I hit a roadblock, the `SpringCamelContext` needs to implement `SmartLifecycle` if the `start` method is to be called (_last_) from `AbstractApplicationContext::refresh` -- but it cannot as the `isAutoStartup` methods in `SmartLifecycle` and `RuntimeConfiguration` are covariant on primitive type (`boolean` and `Boolean`) which is not allowed in [JLS](https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.4.5). That means that we need to go back to listening on `ContextRefreshedEvent` to start CamelContext 😞

davsclaus commented:
The two `isAutoStartup` in spring and camel context (runtime configuration) is likely not 100% the same. The latter controls Camel to whether to start its routes or not on startup, but CamelContext is being started always. I can only assume spring `autoStartup` has a different purpose.

zregvart commented:
Yes, totally different. I'm working on a new strategy to use ContextRefreshed event to start the context and Lifecycle to stop. Running into other issues in camel-spring-boot now.

zregvart commented:
@davsclaus, @nicolaferraro I would very much appreciate if you could have another look, thanks 🥇 

So, two weeks into this, this has not been entirely pain free. I think, now, we have something. The idea is: use `ContextRefreshedEvent` for startup and `Lifecycle::stop` for stoppage, the reason why `Lifecycle` can't be used for startup is that only beans that implement `SmartLifecycle` are started and `SpringCamelContext` can't implement `SmartLifecycle` beacuse of overlapping `isAutoStartupMethod` in `RuntimeConfiguration` and `SmartLifecycle` with different return types -- that cannot be used as covariant return type as the `SmartLifecycle` signature uses primitive type. Using `ContextRefreshedEvent` is actually better than using `SmartLifecycle` as `SmartLifecycle` invocation precedes `ContextRefreshedEvent` emission, so there could be beans in the context that also start on `ContextRefreshedEvent` and Camel would start before those.
One side note on using events is that at some point Spring might default, however unlikely that seems, to using multithreaded event listener invocation, so this might complicate things.
The order of startup is quite important, as some changes to Camel context are not properly expressed if made after context startup (for instance `EndpointStrategy` change), so `SpringCamelContext` has `LOWEST_PRECEDENCE` and `RoutesCollector`s, `CamelContextFactoryBean` have +1 precedence over that in order to add routes and setup context before startup.
During all this tests failed that one would not imagine would have anything to do with this change (such as mock endpoint support in Spring Boot), so this has left me fearful of the possible issues still not uncovered by the tests, please set aside the time to review this thoroughly in order to prevent those. Thank you for taking a look 😺 

Still, I have a few open issues:

 1. is camel-core-xml dependency required? how did it work without it, I was getting ClassNotFoundException when running unit tests (from IDE) (see [components/camel-spring-boot/pom.xml](https://github.com/apache/camel/pull/1685/commits/2fefda07b3e92c14da42f9fbf3bc799b5506d7ae#diff-ec3e60b27e68c04eec43068cb224cc91R87))
 1. are my hypothesises correct (see [hypothesises commit](https://github.com/apache/camel/pull/1685/commits/2fefda07b3e92c14da42f9fbf3bc799b5506d7ae))
 1. there is one quite suspicious change I needed to make for Spring Boot tests (see [dodgy change](https://github.com/apache/camel/pull/1685/commits/074a84215930b8e1937ea6cce1315ad7e499fa61)), I wonder if that is OK, and if there are other places (like in `camel-test-spring`) that I've missed a needed change like that
 1. in the Quartz failover tests I had to change the test itself (see [SpringQuartzConsumerTwoAppsClusteredRecoveryTest](https://github.com/apache/camel/pull/1685/commits/61b3f002bdcc138659b3f8359084cc892d9a2fee#diff-5595d2717727ae3e6a082f06f9eabe34R48) and [SpringQuartzTwoAppsClusteredFailoverTest](https://github.com/apache/camel/pull/1685/commits/61b3f002bdcc138659b3f8359084cc892d9a2fee#diff-d77799eb97ad2117415f07b69c753d4dR63)), it seems weird to me that something like this was not needed before
 1. `CamelContextFactoryBean` and `SpringCamelContext` now implement `Lifecycle`, `Ordered` and `Phased` (see [code changes commit](https://github.com/apache/camel/pull/1685/commits/c93fb42bf81f1da2597660fd546e655acc77597a))
 1. I could have missed a collaborator that also needs to implement `Ordered` and is not reflected in the tests
 1. there is no test for Spring XML configuration (I'm hoping that one with Java configuration of `CamelContextFactoryBean` suffices)
 1. I'm running a full build (I've run tests in \*spring\* components) and I'll post the results later

zregvart commented:
Just a quick note, when running tests from this branch it is best to remove all Camel artefacts (for 2.20.0-SNAPSHOT) from your local Maven repository, in order to eliminate tests being run with an outdated version, It seems that dependencies from Spring Boot -starter jars might be used before the new version is installed in the local Maven repository, encountered when running tests on `camel-spring-cloud`.

zregvart commented:
The full clean build is passing for me :tada:

![](https://i.giphy.com/Za2fXikif3b9e.gif)

davsclaus commented:
We should reach out to the wildfly-camel team and let them test these changes on wildfly as it may affect them also. cc @jamesnetherton @tdiesler

davsclaus commented on components/camel-spring/src/main/java/org/apache/camel/spring/SpringCamelContext.java:
```diff
@@ -119,59 +119,85 @@ public static SpringCamelContext springCamelContext(String configLocations) thro
+    public void start() {
+        // for example from unit testing we want to start Camel later (manually)
+        if (Boolean.TRUE.equals(NO_START.get())) {
+            LOG.trace("Ignoring maybeStart() as NO_START is false");
```
Update the logging message as its not maybeStart anymore

davsclaus commented on components/camel-spring/src/main/java/org/apache/camel/spring/SpringCamelContext.java:
```diff
@@ -119,59 +119,85 @@ public static SpringCamelContext springCamelContext(String configLocations) thro
-            // ContextStoppedEvent is emitted when Spring is end of shutdown
+        } else {
+            // ignore as Camel is already started
+            LOG.trace("Ignoring maybeStart() as Apache Camel is already started");
```
Ditto

davsclaus commented on components/camel-spring/src/main/java/org/apache/camel/spring/SpringCamelContext.java:
```diff
@@ -119,59 +119,85 @@ public static SpringCamelContext springCamelContext(String configLocations) thro
             }
+        } else {
+            // ignore as Camel is already stopped
+            LOG.trace("Ignoring maybeStop() as Apache Camel is already stopped");
```
Ditto

davsclaus replied:
And btw we tend to write Camel instead of Apache Camel in these log messages, maybe we should just do that. Yeah its friday and its 'bike shed' day ;)

davsclaus commented on components/camel-spring/src/main/java/org/apache/camel/spring/SpringCamelContext.java:
```diff
@@ -306,4 +308,24 @@ public String toString() {
+
+    @Override
+    public boolean isRunning() {
+        return isStarted();
```
Just a thought is the contract for running that its fully started? As Camel has a state called .. starting which is the phase it do when it startup. But I assume running is after all the startup stuff and its really running.

zregvart replied:
I think it's fine like this, this would affect Spring lifecycle calling start/stop. So for start we are not using lifecycle (as the code stands now), and for stop we could have just as easily returned `true` and let ServiceSupport handle multiple stop invocations (if any). Perhaps a more correct way would be to return `!isStopping() && !isStopped()`.

davsclaus commented on components/camel-test-spring/src/main/java/org/apache/camel/test/spring/CamelSpringBootExecutionListener.java:
```diff
@@ -31,6 +33,7 @@ public void prepareTestInstance(TestContext testContext) throws Exception {
         LOG.info("@RunWith(CamelSpringBootJUnit4ClassRunner.class) preparing: {}", testContext.getTestClass());
 
         Class<?> testClass = testContext.getTestClass();
+        SpringCamelContext.setNoStart(true);
```
Ah nice way to not start. Maybe add a little bit of code comment that this is on purpose and we do it to ensure the other stuff gets setup before camel is started.

I wonder if the regular camel-test-spring junit classes need anything like this, or do they already do something similar, in case they need to do some preparation and force camel to be started last. Just a thought.

zregvart replied:
Something similar is already done by the Camel Spring Test support (non Spring Boot) in [CamelSpringDelegatingTestContextLoader](https://github.com/apache/camel/blob/83ee9a5e58375e7bec8795e76f57db2fd6b5558c/components/camel-test-spring/src/main/java/org/apache/camel/test/spring/CamelSpringDelegatingTestContextLoader.java#L54-L59), so I think we're OK here.

zregvart commented:
Thanks @davsclaus for reviewing I've added clarifications and made that `isRunning` method a bit more correct in implementation. I'll run full tests once again and leave some time for further review before I squash this and merge into master.

zregvart commented:
I've merged this to `master`, let's try to get some feedback before evaluating if this would be merged to `camel-2.19.x` also.

davsclaus commented:
Yeah we can think about a backport.

Do you mind add a note about this change  to the release notes
http://camel.apache.org/camel-2200-release.html
